Source/Core/Mock.jsl

```diff
@@ -0,0 +1,225 @@
+// instrumented function, its expectations, and its current expectation. Part of the class is here and the rest is
+// added dynamically by ut mock function. A class is used rather than a namespace so it cleans up after itself.
+Define Class("UtMockFunction",
+	n calls = 0;
+	expectations = {};
+);
+
+/* 
+	Function: ut mock function
+		---Prototype---
+		ut mock function( {args}, <{locals}>, Expr(body) )
+		---------------
+		Used to test expectations on callbacks and the like.
+
+		Creates a mock function that creates a function from the pieces provided and decorates it.
+		Returns an object (called obj in this document) with the decorated function accessible via obj:fn.
+		You can then use ut expect call(obj, ...) with a list of matchers to set expectations. You should call
+		this once for each time you expect your function to be called. This should be done before the calls to
+		your function. As your function is called, these expectations are checked one by one. Then, after you
+		test body is over, call ut verify mock(obj). This will make sure you got the correct number of calls.
+
+		Note, it is /extremely/ important that you ensure that ut verify mock is called. Otherwise, your test
+		will *silently pass* even if you didn't get all the calls you expected. The easiest way to ensure this
+		is to call it in the Teardown of a ut test case.
+
+		Users of other mock frameworks may notice that this is a little more than a mock since we are
+		instrumenting an arbitrary function. However, this was the easiest way to supply both the signature and
+		return to the ut mock function factory. And it gives us some extra flexibility. So hopefully you'll
+		forgive me stretching the definition. Also, it doesn't have the capability (yet) to change the body
+		of the function for each expectation. However, this is something I hope to add.
+
+		Another difference from other mock frameworks is that this verifies the expectations eagerly in the
+		instrumented function. This is useful when the arguments may not live past the callback call and they
+		can not be cloned. One common example of this is Display Boxes.
+
+		Use obj:fn(...) to call the instrumented function. And use obj:orig fn(...) to call the uninstrumented
+		function.
+		
+		When calling ut mock function, you can omit the locals list if desired.
+
+	Arguments:
+		args - args list
+		locals - locals list
+		body - an expression using args and locals 
+
+	Example:
+		---JSL---
+		Button Box Test = ut test case("Button Box")
+		    << Setup(Expr(mock = ut mock function({this}, Expr(.))))
+		    << Teardown(Expr(ut verify mock(mock)));
+		ut test(Button Box Test, "Calls Callback", Expr(
+		    ut expect call(mock, {ut enabled(1)});
+		    New Window("", btn = Button Box("Hi", << Set Function(Function({this}, mock:fn(this)))));
+		    btn << Click;
+		));
+		---------
+*/
+ut mock function = Expr(Function({fnarg1, fnarg2, fnarg3="NOT PROVIDED"}, {obj, body},
```

Comment:
Not saying this is bad, but all of our other documented functions put the full definition in the call to `ut define documented function`.  The expr would get overwritten in the function call later on.